### PR TITLE
Add option to check dataset labels in SFTTrainer

### PR DESCRIPTION
**Pull request by geronimi73**
Hi everyone!

This PR introduces a new option `check_dataset_labels` in `SFTTrainer`. When enabled, the trainer calls the collator on the first sample from the training set and logs `token_id`, decoded `token_id`, and the corresponding `label`. This helps to uncover mistakes early, such as setting the tokenizer's `pad_token` to`eos_token`. The idea is taken from axolotl where such an option exists already. 

**Problem(s)**:
It's common practice to set the tokenizer's pad_token to eos_token. This is problematic because `DataCollatorForCompletionOnlyLM` sets the `label` for all occurrences of `pad_token` to `-100`. If `pad=eos`, then **the model will never learn to output eos**. This issue can be hard to debug and many people struggle with it.

Additionally, when using `instruction_template` and `response_template`, logging the tokens and labels would be helpful to ensure that all the labels are correctly set to only train on output and ignore the instruction. Furthermore, tokenizers can be complex and the output would aid in spotting tokenization issues like handling of special tokens quickly

**Solution**:
* Add a new option `check_dataset_labels` to the trainer.
* When check_dataset_labels is enabled, the trainer calls the collator on the first sample from the training set and logs token_id, decoded token_id, and the corresponding label.
* This creates transparency and helps to uncover mistakes early.

**Usage**:
```python
# load model and tokenizer 
...

messages = [
    {"role": "user", "content": "Hello who are you?"},
    {"role": "assistant", "content": "Luke, I am your father"},
    {"role": "user", "content": "WTF"},
]
dataset = Dataset.from_list([dict(messages=messages)])

trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    data_collator = DataCollatorForCompletionOnlyLM(
        instruction_template = "<|im_start|>user", 
        response_template = "<|im_start|>assistant", 
        tokenizer = tokenizer, 
        mlm = False),
    check_dataset_labels = True,
    dataset_kwargs=dict(add_special_tokens=False),
    args = TrainingArguments(output_dir = "out")
)
```

output:
```
check_dataset_labels:
<|im_start|> user
Hello who are you? <|im_end|> 
 <|im_start|> assistant
Luke, I am your father <|im_end|> 
 <|im_start|> user
WTF <|im_end|> 

32000 '<|im_start|>' -100
1792 'user' -100
13 '<0x0A>' -100
10994 'Hello' -100
1058 'who' -100
526 'are' -100
366 'you' -100
29973 '?' -100
32001 '<|im_end|>' -100
13 '<0x0A>' -100
32000 '<|im_start|>' -100
465 'ass' -100
22137 'istant' -100
13 '<0x0A>' 13
24126 'Lu' 24126
446 'ke' 446
29892 ',' 29892
306 'I' 306
626 'am' 626
596 'your' 596
4783 'father' 4783
32001 '<|im_end|>' 32001
13 '<0x0A>' 13
32000 '<|im_start|>' -100
1792 'user' -100
13 '<0x0A>' -100
29956 'W' -100
8969 'TF' -100
32001 '<|im_end|>' -100
13 '<0x0A>' -100
```

**Related issues**:
* https://github.com/huggingface/trl/issues/1384
* https://github.com/huggingface/trl/issues/976
* https://github.com/huggingface/trl/issues/1191
* https://github.com/huggingface/trl/issues/916
* https://github.com/huggingface/transformers/issues/22794
* https://stackoverflow.com/questions/76446228/setting-padding-token-as-eos-token-when-using-datacollatorforlanguagemodeling-fr

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Amazing work ! 
Can you add the docstring of that arg in SFTTrainer's docstring together with few lines explaining what you posted on the PR? 🙏

**Review by younesbelkada — COMMENTED** [review]
Thanks @geronimi73 
the changes look great ! can you just run the styling checks? `make precommit` , then we can merge

**Comment by geronimi73** [comment]
> Thanks @geronimi73 the changes look great ! can you just run the styling checks? `make precommit` , then we can merge

what about the `print` statements I mentioned? leave it like this?

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -302,6 +305,15 @@ def make_inputs_require_grad(module, input, output):
...
             )
 
+        if check_dataset_labels:
+            if train_dataset is not None and len(train_dataset) > 0:
+                input_ids, attention_mask, labels = data_collator([train_dataset[0]]).values()
+
+                print("check_dataset_labels:")
+                print(tokenizer.decode(input_ids[0]))
+                for token, label in zip(input_ids[0], labels[0]):
+                    print(f"{token.item()}, '{tokenizer.decode(token)}' {label.item()}")
```
Comment (younesbelkada) [inline trl/trainer/sft_trainer.py:315]:
```suggestion
                print("check_dataset_labels:") # noqa
                print(tokenizer.decode(input_ids[0])) # noqa
                for token, label in zip(input_ids[0], labels[0]):
                    print(f"{token.item()}, '{tokenizer.decode(token)}' {label.item()}") # noqa
```

**Review by younesbelkada — COMMENTED** [review]
Yes I think we can leave the print

**Review by younesbelkada — COMMENTED** [review]
Hi @geronimi73 !
Again thank you very much for this contribution, after thinking a bit, I think we should make things standardized and use things like `logger.info` instead of `print`, so that it approach is harmonized across our coding practice in the HF codebase - sorry for that as you already asked the question and i said yes! - would you mind switching `print` statements to `logger.info` ? you will need to import `logging` at the top of the file and import the logger properly ! let me know if you need any help or if you think we should keep the print statements

**Comment by younesbelkada** [comment]
(to fix the failing tests you just have to rebase with main)

**Comment by geronimi73** [comment]
> you will need to import `logging` at the top of the file and import the logger properly

i tried. it works but not sure if this is the way to do it. please check my comment on the code

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -302,6 +307,18 @@ def make_inputs_require_grad(module, input, output):
                 "overflow issues when training a model in half-precision. You might consider adding `tokenizer.padding_side = 'right'` to your code."
             )
 
+        # set the correct log level depending on the node
+        log_level = args.get_process_log_level()
+        logger.setLevel(log_level)
```
Comment (younesbelkada) [inline trl/trainer/sft_trainer.py:312]:
```suggestion
```

Comment (younesbelkada) [reply trl/trainer/sft_trainer.py:312]: I think this piece might be not needed

**Review by younesbelkada — COMMENTED** [review]
Thanks ! I left one comment ! can you also run the styling checks? `make precommit`

**Comment by younesbelkada** [comment]
Hi @geronimi73 sorry for all the iteration ! can you re-run the styling checks again? 🙏

**Comment by geronimi73** [comment]
> Hi @geronimi73 sorry for all the iteration ! can you re-run the styling checks again? 🙏

sure! but i'm still wondering whether it works. I never see the output of `logger.info()`!

how am I supposed to set the log level (as a user of SFTTrainer) ?  I tried this, doesnt work:
```python
import transformers
transformers.logging.set_verbosity_info()
```

**Comment by geronimi73** [comment]
> > Hi @geronimi73 sorry for all the iteration ! can you re-run the styling checks again? 🙏
> 
> sure! but i'm still wondering whether it works. I never see the output of `logger.info()`!
> 
> how am I supposed to set the log level (as a user of SFTTrainer) ? I tried this, doesnt work:
> 
> ```python
> import transformers
> transformers.logging.set_verbosity_info()
> ```

also this one does not enable it 
```python
args = TrainingArguments(
    output_dir = "out",
    log_level = "info"
    )
```

**Comment by geronimi73** [comment]
sorry to keep bothering you @younesbelkada but I think the problem ist that we are using logging from `transformers.utils` and obtaining a logger with `__name__` (=`trl.trainer.sft_trainer`). This means that setting the log_level in `TrainingArguments` does not have an effect on the logger we obtained because [this line](https://github.com/huggingface/transformers/blob/092f1fdaa4224fdd88c616dc9678e6fcb37bfffd/src/transformers/trainer.py#L375) in `trainer.py` affects the root logger `transformers` and does not have any effect on our logger (`trl.trainer.sft_trainer`).

To do this correctly I think, we would have to either 
* add a `trl.utils.logger` analogous to `transformers.utils.logger` (too much effort?) 
* OR use `logger.setLevel(log_level)` in `SFTTrainer.init()`. Otherwise we will not see the output of `logging.info()`.

what do you think? option 2?
